source/common/stats/stats_impl.h

```diff
@@ -473,5 +477,131 @@ class IsolatedStoreImpl : public Store {
   IsolatedStatsCache<Histogram, HistogramImpl> histograms_;
 };
 
+/**
+ * Structure to hold statistical data of histogram.
+ */
+struct HistogramStatistics {
+public:
+  HistogramStatistics(const std::string name, histogram_t* histogram_ptr) : name_(name) {
+    double arr[9] = {0, 0.25, 0.5, 0.75, 0.90, 0.95, 0.99, 0.999, 1};
+    double out[9];
+    double* arrptr = arr;
+    double* outptr = out;
+    hist_approx_quantile(histogram_ptr, arrptr, 9, outptr);
+    p0_ = out[0];
+    p25_ = out[1];
+    p50_ = out[2];
+    p75_ = out[3];
+    p90_ = out[4];
+    p95_ = out[5];
+    p99_ = out[6];
+    p999_ = out[7];
+    p100_ = out[8];
+  }
+
+  std::string name() const { return name_; }
+
+  std::string summary() const {
+    return fmt::format("P0: {} , P25: {}, P50: {}, P75: {}, P90: {}, P95: {}, P99: {}, P100: {}",
+                       p0_, p25_, p50_, p75_, p90_, p95_, p99_, p999_, p100_);
+  }
+
+private:
+  const std::string name_;
+  double p0_;
+  double p25_;
+  double p50_;
+  double p75_;
+  double p90_;
+  double p95_;
+  double p99_;
+  double p999_;
+  double p100_;
+};
+
+/**
+ * BaseSink implementation for all Sinks. This mainly  .
+ */
+class BaseSink : public Sink {
+public:
+  void onHistogramComplete(const Histogram& hist, uint64_t value) override {
+    ThreadLocalHistograms tls_histograms = hist_tls_->getTyped<ThreadLocalHistograms>();
+    auto tls_hist_iterator = tls_histograms.histograms_.find(hist.name());
+    if (tls_hist_iterator != tls_histograms.histograms_.end()) {
+      hist_insert(tls_hist_iterator->second, value, 1);
+    } else {
+      histogram_t* histogram_ptr = hist_alloc();
+      hist_insert(histogram_ptr, value, 1);
+      tls_histograms.histograms_.emplace(std::make_pair(hist.name(), histogram_ptr));
+    }
+  }
+
+  std::list<HistogramStatistics> flushHistograms() override {
+    // TODO: Merge with optimized locking
+    std::lock_guard<std::mutex> guard(lock_);
+    hist_tls_->runOnAllThreads([this]() {
```

Comment:
The code as written I don't think works. `runOnAllThreads()` will return immediately and all the thread functions will run asynchronously.

This unfortunately is what I was saying about this being pretty complicated. At a high level, here is the optimal solution:
1) I would rework the stats allocation code so that a `Histogram` is a first class thing just look a `Counter` and a `Gauge`. This will fix various other issues about extracting current values in admin dumping, etc. It will also handle the registration part so that you don't have to register on first use and dynamically build a merge histogram.
2) Flushing stats becomes a multi-stage process. First, start flush on central thread. Launch flush on each TLS worker. On each TLS worker, swap histograms that are being written to (hist1 <-> hist2) . When all workers have swapped, post back to main thread, and finish flush. Now all merging can be done without any locks as you know no one is writing to them.

If you don't want to do the above, I would prefer you actually just do this for now in a dedicated Prometheus/metrics service stat sink and then someone can come along and do a more optimal/general solution later. Still, you will need to actually build a barrier below and wait for all threads to check in before continuing, and have proper locking.

Reply:
i will first implement this properly (with barrier and locking) at BaseSink enabling only for selected sinks (Metrics). After that we can take a call about other sinks. For barrier implementation, is there a way to get number of worker threads?